File: HW/HW11/Solution.py
from typing import List


# Recursive Approach
def floodFill(image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
    original_color = image[sr][sc]
    if original_color == color:
        return image

    def dfs(r, c):
        if image[r][c] == original_color:
            image[r][c] = color
            if r >= 1:
                dfs(r - 1, c)
            if r + 1 < len(image):
                dfs(r + 1, c)
            if c >= 1:
                dfs(r, c - 1)
            if c + 1 < len(image[0]):
                dfs(r, c + 1)

    dfs(sr, sc)
    return image


# Iterative Approach
from typing import List
from collections import deque


def floodFill(
    image: List[List[int]], sr: int, sc: int, newColor: int
) -> List[List[int]]:
    original_color = image[sr][sc]
    if original_color == newColor:
        return image

    rows, cols = len(image), len(image[0])
    queue = deque([(sr, sc)])

    while queue:
        r, c = queue.popleft()
        if image[r][c] == original_color:
            image[r][c] = newColor
            if r >= 1:
                queue.append((r - 1, c))
            if r + 1 < rows:
                queue.append((r + 1, c))
            if c >= 1:
                queue.append((r, c - 1))
            if c + 1 < cols:
                queue.append((r, c + 1))

    return image

File: HW/HW11/Solution.java
package HW.HW11;

import java.util.LinkedList;
import java.util.Queue;

// Recursive Approach
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int originalColor = image[sr][sc];
        if (originalColor == color) {
            return image;
        }

        floodFillHelper(image, sr, sc, color, originalColor);
        return image;
    }

    private void floodFillHelper(int[][] image, int sr, int sc, int color, int originalColor) {
        if (sr < 0 || sr >= image.length || sc < 0 || sc >= image[0].length || image[sr][sc] != originalColor) {
            return;
        }

        image[sr][sc] = color;

        floodFillHelper(image, sr - 1, sc, color, originalColor); // up
        floodFillHelper(image, sr + 1, sc, color, originalColor); // down
        floodFillHelper(image, sr, sc - 1, color, originalColor); // left
        floodFillHelper(image, sr, sc + 1, color, originalColor); // right
    }
}

// Iterative Approach
class Solution2 {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int originalColor = image[sr][sc];
        if (originalColor == newColor) {
            return image;
        }

        int rows = image.length, cols = image[0].length;
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[] { sr, sc });

        while (!queue.isEmpty()) {
            int[] cell = queue.poll();
            int r = cell[0], c = cell[1];

            if (image[r][c] == originalColor) {
                image[r][c] = newColor;
                if (r >= 1)
                    queue.add(new int[] { r - 1, c });
                if (r + 1 < rows)
                    queue.add(new int[] { r + 1, c });
                if (c >= 1)
                    queue.add(new int[] { r, c - 1 });
                if (c + 1 < cols)
                    queue.add(new int[] { r, c + 1 });
            }
        }

        return image;
    }
}
